### ISSUE TITLE
Create an abstraction over the metadata source

### ISSUE THREAD

**Issue by slrtbtfs** [prometheus-community/promql-langserver#134]
Currently metadata is retrieved by having a lot of prometheus API calls spread over the language server.

It would be better to have a interface roughly like this:

```
type MetadataSource interface {
      Labels() []string
      LabelValues(label string) []string
      ....
}
```

Then multiple implementations can be provided.

* One that just forwards the calls to a the v1 API as it is done now.
* A similar one but with caching for better performance. #66 
* One that does not use the network stack but direcly communicates with a prometheus v1 API object. #117 
* Implementations that use Thanos or Cortex as a data source.

**Comment by haibeey** [comment]
this relies on [#177](https://github.com/prometheus-community/promql-langserver/issues/117)  I guess ????

**Comment by slrtbtfs** [comment]
The other way around.

This one is the preparation for #177.

**Comment by Nexucis** [comment]
I think I will do the first and the third implementation for the moment.

I guess the caching one should wrappe the others and should not be a complete new implementation.

Let's see for the last one

**Comment by slrtbtfs** [comment]
That sounds reasonable.

**Comment by Nexucis** [comment]
After looking at the code, I'm feeling maybe it would be a better idea to build a sort of prometheus client that would respect these interface :

```go
type PrometheusClient interface {
   Version() string
   IsCompatible() bool
   Labels() []string
   LabelValues(label string) []string
}
```
I'm not yet sure about the name of the interface.

But the idea using this kind of client will be to remove the usage of the boolean `headless` and basically to simplify the code of the server by delegating the logic of how to retrieve the information to the client.
Basically it will be the client that will tell to the langueserver if it is compatible and so on.

I'm also pretty sure we can remove the usage of the mutex `prometheusMu`.

If you are agree with this approach, then I think it will make sense that the http client is indeed implemented here because it is required by lsp. 

But I think it will make sense that the client that doesn't use the network should be implemented on Prometheus side directly. Like that it delegates the responsability of who is doing what.

What do you think about this approach?

**Comment by Nexucis** [comment]
it's more or less what you already proposed, it is just an enhancement of the initial idea :)

**Comment by Nexucis** [comment]
Finally for Thanos, we just need to expose the rest api directly in thanos like you started to do in Prometheus.
And so if needed we will have to code the prometheusClient in thanos directly ( I believe it requires the query agregation that thanos is doing, so it's really something dedicated to the business of thanos and not related to this project)

**Comment by slrtbtfs** [comment]
> What do you think about this approach?

Look good in general. Some minor comments though:

>    IsCompatible() bool

The purpose of this is special casing about whether the Prometheus instance supports the new metadata API. Ideally this special casing would not be exposed by the interface but the implementation would emulate the interface if it is not present.

> I'm also pretty sure we can remove the usage of the mutex prometheusMu.

We could add a `ChangeDataSource(url string) error` function to the interface. On implementations that talk to a remote Prometheus this would change the url which would need to be protected by an mutex inside the interface implementation. On non remote implementations it would just always fail.

> If you are agree with this approach, then I think it will make sense that the http client is indeed implemented here because it is required by lsp.
>
> But I think it will make sense that the client that doesn't use the network should be implemented on Prometheus side directly. Like that it delegates the responsability of who is doing what.
Maybe for this last idea, @juliusv or @roidelapluie would have some other things in mind ?

Sounds good.

> Finally for Thanos, we just need to expose the rest api directly in thanos like you started to do in Prometheus.

It seems to be already implemented in some way: https://thanos.io/components/query.md/

**Comment by Nexucis** [comment]
> >IsCompatible() bool

> The purpose of this is special casing about whether the Prometheus instance supports the new metadata API. Ideally this special casing would not be exposed by the interface but the implementation would emulate the interface if it is not present.

Ah yeah, that's a better idea indeed. Let's see if it's feasible :)

>>I'm also pretty sure we can remove the usage of the mutex prometheusMu.

> We could add a ChangeDataSource(url string) error function to the interface. On implementations that talk to a remote Prometheus this would change the url which would need to be protected by an mutex inside the interface implementation. On non remote implementations it would just always fail.

Good idea, thanks !

>> Finally for Thanos, we just need to expose the rest api directly in thanos like you started to do in Prometheus.

> It seems to be already implemented in some way: https://thanos.io/components/query.md/

mmm since thanos is following the same rest interface than prometheus, then we will have to add the endpoint /completion, /diagnostics ...etc. Well if they are agree of course

**Comment by slrtbtfs** [comment]
> mmm since thanos is following the same rest interface than prometheus, then we will have to add the endpoint /completion, /diagnostics ...etc. Well if they are agree of course

Let's get it working with vanilla Prometheus first.

**Comment by slrtbtfs** [comment]
For using Prometheus directly without a network stack I think it would be best to have another `Client` implementation which is constructed from an `API` object from https://github.com/prometheus/prometheus/blob/000ba352779e3b0597ed19129fa90c3399257a03/web/api/v1/api.go#L172 .

Since including the `prometheus/web/api/v1` package would pull in a lot of dependencies, that would best be implemented in a package that is not imported by the normal langserver binary.

**Comment by Nexucis** [comment]
I didn't check yet how to do it without thé network, but for me I'm still thinking it would be for the best that the implementation is on prometheus side.
Like you said it will require many things inside Prometheus to make it works and it's still a pain to have prometheus as a dependancy even if with the go module it has been improved

But for that I can't take the decision. Someone from Prometheus would have to be agree on that.

**Comment by Nexucis** [comment]
Looking at the code you pointed, I'm not sure it's a good approach to use the API object since it is dedicated to initialiaze the http endpoints. Using it to request Prometheus internaly would be a bit weird.
But we can certainly take some inspirations on this code to build the internal client :)

**Comment by slrtbtfs** [comment]
The language already has the Prometheus module as a dependency. 

But the implementation can live inside the Prometheus sources, as well. If it is ok for the Prometheus maintainers, the `API` object could directly implement the language server interface.

> And in any case, I will be able to implement it if you need more help :)

That would be great! Are you around on IRC, so we could discuss this a bit more synchronously?

**Comment by slrtbtfs** [comment]
> Looking at the code you pointed, I'm not sure it's a good approach to use the API object since it is dedicated to initialiaze the http endpoints. Using it to request Prometheus internaly would be a bit weird.

True.

**Comment by Nexucis** [comment]
> Are you around on IRC, so we could discuss this a bit more synchronously?
Not currently, but I can go on it. But maybe more tomorrow :)

**Comment by Nexucis** [comment]
I ping you on IRC but not sure you receive it. It's been a while I didn't go on it. I certainly missed a step :)

**Comment by slrtbtfs** [comment]
I think this issue can be closed, since the abstraction exists now.

Individual implementations of the `prometheus.Client` interface are already tracked in separate issues.
